### ISSUE TITLE
incorrect Git completions: absolute paths (starting with `:`) aren’t properly escaped

### ISSUE THREAD

**Issue by ariasuni** [fish-shell/fish-shell#9004]
fish 3.4.1

How to reproduce:
1. Create a new Git repository (or use an existing one)
2. Create a file named `_test`, and a directory called `dir`
3. Go to `dir`
4. Type `git add`, hit `<TAB>`, it suggests `:/_test`
5. Hit enter, you get `fatal: Unimplemented pathspec magic '_' in ':/_test`

According to https://git-scm.com/docs/gitglossary#Documentation/gitglossary.txt-aiddefpathspecapathspec:

> A pathspec that begins with a colon `:` has special meaning. In the short form, the leading colon `:` is followed by zero or more "magic signature" letters (which optionally is terminated by another colon `:`), and the remainder is the pattern to match against the path. The "magic signature" consists of ASCII symbols that are neither alphanumeric, glob, regex special characters nor colon. The optional colon that terminates the "magic signature" can be omitted if the pattern begins with a character that does not belong to "magic signature" symbol set and is not a colon.

Here the magic signature `:/` is used, but since `_` is «neither alphanumeric, glob, regex special characters nor colon», so it thinks `_` has a special meaning. Instead, one can escape such characters (e.g. `:/\_test`), or end the magic signature so that the rest is interpreted literally (e.g. `:/:_test`).

It feels like the second one is easy to implement and will avoid other such ambiguity or errors.

**Comment by faho** [comment]
Hmmm... I'd assume it's probably easier to just terminate the magic signature bit - make it `:/:test` instead of `:/test`.

**Comment by zanchey** [comment]
This broke the tests on CentOS 7 (which has a fairly old version of Git), though I can't immediately see why:

```
[  316s] 140/236 Test #132: git.fish .................................***Failed    0.86 sec
[  316s] Testing file checks/git.fish ... [31mFailure[0m:
[  316s] 
[  316s]   The CHECK on line 44 wants:
[  316s]     [1m:/:foo\tUntracked file[0m
[  316s] 
[  316s]   which failed to match line stdout:3:
[  316s]     [1m (newbranch)[0m
[  316s] 
[  316s]   Context:
[  316s]     [1madd\tAdd file contents to the index[0m
[  316s]     [1mfoo\tUntracked file[0m
[  316s]      <= nothing to match CHECK '[34m:/:foo\tUntracked file[0m' on line 44
[  316s]     [1m (newbranch)[0m
[  316s]     [1m (newbranch|âœ”)[0m
[  316s]     [1m (newbranch|â€¦1)[0m
[  316s] 
[  316s]   when running command:
[  316s]     ../test/root/bin/fish -i checks/git.fish
```

**Comment by faho** [comment]
Ah, lovely. CentOS 7's git is what I would call "comically old". In fact it's so old it's older than the git homepage hosts docs for (they go back to 2.12.5 from 2017, CentOS 7 has 1.8.3 from 2015).

So apparently it doesn't support the `:/:` style. Easy enough to fix, we already check a git version here, we can simply keep the correction only in the new path. (unfortunately we can't really do the escaping here because we'd have to enumerate the characters, and the git docs don't actually list them all - if that version can even do that)

(alternatively we can disable the git tests if git is *that* old)

**Comment by zanchey** [comment]
It should work, according to the commit history. https://github.com/git/git/commit/8a42c9850177cc91e9f38779e8aca89682a02975 says that the short form is supported, but perhaps there was a bug?

**Comment by faho** [comment]
@zanchey Do you have access to a CentOS 7 machine?

Can you confirm that e.g. creating a file called `foo` in the repo root and adding it with `git add :/:foo` doesn't work but adding it with `git add :/foo` does?

Honestly I'm really leaning towards just disabling the tests for git that old - we don't know what bugs it has and we'd have to make the tests for the new versions *worse* to test it too (e.g. matching `:/{{:?}}foo` in littlecheck parlance).

**Comment by zanchey** [comment]
My emailed reply disappeared?



> Can you confirm that e.g. creating a file called `foo` in the repo root and adding it with `git add :/:foo` doesn't work but adding it with `git add :/foo` does?

No, both of these things appear to work. I'll dig into it a bit more.

**Comment by zanchey** [comment]
OK, I was getting buried in the v2 porcelain stuff, but of course git this old only supports porcelain v1. The problem is that this branch calculates the ':/:'-prefixed path, but then because the path is valid as a relative path, wipes it out again.

The following patch seems to fix it for me:
```patch
diff --git a/share/completions/git.fish b/share/completions/git.fish
index 35cd87faa..92fad633d 100644
--- a/share/completions/git.fish
+++ b/share/completions/git.fish
@@ -487,6 +487,7 @@ function __fish_git_files
                 end
 
                 if test "$root/$relfile" = (pwd -P)/$relfile
+                    and not string match -q ':*' -- (commandline -ct)
                     set file $relfile
                 end
 
```

Does that look sensible?

**Comment by krobelus** [comment]
yep that looks good.

**Comment by zanchey** [comment]
Merged in 68f67ff46cd4f10cb8f770199cee755c34397758

**Comment by zanchey** [comment]
> Can you confirm that e.g. creating a file called `foo` in the repo root 
> and adding it with `git add :/:foo` doesn't work but adding it with `git 
> add :/foo` does?

Actually, it does work fine.

systemd-nspawn has been useful for me to get these distributions running 
in a container.
